File: Module/PlayWays/src/main/java-songmanage/com/module/playways/songmanager/adapter/RecommendSongAdapter.java
package com.module.playways.songmanager.adapter;

import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.common.utils.U;
import com.common.view.DebounceViewClickListener;
import com.common.view.ex.ExTextView;
import com.common.view.ex.drawable.DrawableCreator;
import com.common.view.recyclerview.DiffAdapter;
import com.common.view.recyclerview.RecyclerOnItemClickListener;
import com.module.playways.R;
import com.module.playways.room.song.model.SongModel;
import com.zq.live.proto.Common.StandPlayType;

public class RecommendSongAdapter extends DiffAdapter<SongModel, RecyclerView.ViewHolder> {

    boolean isOwner;
    RecyclerOnItemClickListener<SongModel> mListener;

    public RecommendSongAdapter(boolean isOwner, RecyclerOnItemClickListener<SongModel> l) {
        this.isOwner = isOwner;
        this.mListener = l;
    }

    Drawable pk = new DrawableCreator.Builder()
            .setStrokeColor(U.getColor(R.color.white_trans_70))
            .setStrokeWidth(U.getDisplayUtils().dip2px(1.5f))
            .setCornersRadius(U.getDisplayUtils().dip2px(10))
            .setSolidColor(Color.parseColor("#CB5883"))
            .build();

    Drawable togather = new DrawableCreator.Builder()
            .setStrokeColor(U.getColor(R.color.white_trans_70))
            .setStrokeWidth(U.getDisplayUtils().dip2px(1.5f))
            .setCornersRadius(U.getDisplayUtils().dip2px(10))
            .setSolidColor(Color.parseColor("#7088FF"))
            .build();

    Drawable game = new DrawableCreator.Builder()
            .setSolidColor(Color.parseColor("#61B14F"))
            .setCornersRadius(U.getDisplayUtils().dip2px(10))
            .setStrokeWidth(U.getDisplayUtils().dip2px(1.5f))
            .setStrokeColor(U.getColor(R.color.white_trans_70))
            .build();

    Drawable freeMic = new DrawableCreator.Builder()
            .setSolidColor(Color.parseColor("#C856E0"))
            .setCornersRadius(U.getDisplayUtils().dip2px(10))
            .setStrokeWidth(U.getDisplayUtils().dip2px(1.5f))
            .setStrokeColor(U.getColor(R.color.white_trans_70))
            .build();

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.recommend_song_item_layout, parent, false);
        ItemHolder viewHolder = new ItemHolder(view, isOwner);
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        SongModel model = mDataList.get(position);

        ItemHolder reportItemHolder = (ItemHolder) holder;
        reportItemHolder.bind(model, position);
    }

    @Override
    public int getItemCount() {
        return mDataList.size();
    }

    private class ItemHolder extends RecyclerView.ViewHolder {
        ExTextView mSongNameTv;
        ExTextView mSelectTv;
        ExTextView mSongTag;
        TextView mSongDesc;

        SongModel mSongModel;

        public ItemHolder(View itemView, boolean isOwner) {
            super(itemView);
            mSongNameTv = (ExTextView) itemView.findViewById(R.id.song_name_tv);
            mSelectTv = (ExTextView) itemView.findViewById(R.id.select_tv);
            mSongTag = (ExTextView) itemView.findViewById(R.id.song_tag);
            mSongDesc = (TextView) itemView.findViewById(R.id.song_desc);

            if (isOwner) {
                mSelectTv.setText("点歌");
            } else {
                mSelectTv.setText("想唱");
            }

            mSelectTv.setOnClickListener(new DebounceViewClickListener() {
                @Override
                public void clickValid(View v) {
                    if (mListener != null) {
                        mListener.onItemClicked(v, -1, mSongModel);
                    }
                }
            });
        }

        public void bind(SongModel model, int position) {
            mSongModel = model;
            mSongNameTv.setText("《" + model.getDisplaySongName() + "》");
            mSongTag.setVisibility(View.VISIBLE);
            if (TextUtils.isEmpty(mSongModel.getSongDesc())) {
                mSongDesc.setVisibility(View.GONE);
            } else {
                mSongDesc.setVisibility(View.VISIBLE);
                mSongDesc.setText(mSongModel.getSongDesc());
            }
            if (model.getPlayType() == StandPlayType.PT_SPK_TYPE.getValue()) {
                mSongTag.setBackground(pk);
                mSongTag.setText("PK");
            } else if (model.getPlayType() == StandPlayType.PT_CHO_TYPE.getValue()) {
                mSongTag.setBackground(togather);
                mSongTag.setText("合唱");
            } else if (model.getPlayType() == StandPlayType.PT_MINI_GAME_TYPE.getValue()) {
                mSongTag.setBackground(game);
                mSongTag.setText("双人游戏");
            } else if (model.getPlayType() == StandPlayType.PT_FREE_MICRO.getValue()) {
                ;
                mSongTag.setBackground(freeMic);
                mSongTag.setText("多人游戏");
            } else {
                mSongTag.setVisibility(View.GONE);
            }
        }
    }
}

File: Module/PlayWays/src/main/java-songmanage/com/module/playways/songmanager/adapter/RecommendSongAdapter.kt
package com.module.playways.songmanager.adapter

import android.support.v7.widget.RecyclerView
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.common.log.MyLog

import com.common.view.DebounceViewClickListener
import com.common.view.ex.ExTextView
import com.common.view.recyclerview.DiffAdapter
import com.common.view.recyclerview.RecyclerOnItemClickListener
import com.module.playways.R
import com.module.playways.room.song.model.SongModel
import com.module.playways.songmanager.SongManagerActivity
import com.module.playways.songmanager.utils.SongTagDrawableUtils
import com.zq.live.proto.Common.StandPlayType

class RecommendSongAdapter(internal var isOwner: Boolean, var type: Int, internal var mListener: RecyclerOnItemClickListener<SongModel>?) : DiffAdapter<SongModel, RecyclerView.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.recommend_song_item_layout, parent, false)
        return ItemHolder(view, isOwner, type)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val model = mDataList[position]

        val reportItemHolder = holder as ItemHolder
        reportItemHolder.bind(model, position)
    }

    override fun getItemCount(): Int {
        return mDataList.size
    }

    private inner class ItemHolder(itemView: View, isOwner: Boolean, type: Int) : RecyclerView.ViewHolder(itemView) {

        var mSelectTv: ExTextView = itemView.findViewById(R.id.select_tv)
        var mSongTag: ExTextView = itemView.findViewById(R.id.song_tag)
        var mSongDesc: TextView = itemView.findViewById(R.id.song_desc)
        var mSongNameTv: ExTextView = itemView.findViewById(R.id.song_name_tv)

        private var mSongModel: SongModel? = null

        init {
            mSelectTv.setOnClickListener(object : DebounceViewClickListener() {
                override fun clickValid(v: View) {
                    if (mListener != null) {
                        mListener!!.onItemClicked(v, -1, mSongModel)
                    }
                }
            })
        }

        fun bind(model: SongModel, position: Int) {
            mSongModel = model
            if (type == SongManagerActivity.TYPE_FROM_MIC) {
                MyLog.d(TAG, "bind ItemHolder = type = $type")
                if (model.playType == StandPlayType.PT_COMMON_TYPE.value) {
                    mSelectTv.text = "点歌"
                } else {
                    mSelectTv.text = "发起"
                }
            } else {
                if (isOwner) {
                    mSelectTv.text = "点歌"
                } else {
                    mSelectTv.text = "想唱"
                }
            }
            mSongNameTv.text = "《" + model.displaySongName + "》"
            mSongTag.visibility = View.VISIBLE
            if (TextUtils.isEmpty(mSongModel?.songDesc)) {
                mSongDesc.visibility = View.GONE
            } else {
                mSongDesc.visibility = View.VISIBLE
                mSongDesc.text = mSongModel?.songDesc
            }
            when {
                model.playType == StandPlayType.PT_SPK_TYPE.value -> {
                    mSongTag.background = SongTagDrawableUtils.pkDrawable
                    mSongTag.text = "PK"
                }
                model.playType == StandPlayType.PT_CHO_TYPE.value -> {
                    mSongTag.background = SongTagDrawableUtils.chorusDrawable
                    mSongTag.text = "合唱"
                }
                model.playType == StandPlayType.PT_MINI_GAME_TYPE.value -> {
                    mSongTag.background = SongTagDrawableUtils.miniGameDrawable
                    mSongTag.text = "双人游戏"
                }
                model.playType == StandPlayType.PT_FREE_MICRO.value -> {
                    mSongTag.background = SongTagDrawableUtils.freeMicDrawable
                    mSongTag.text = "多人游戏"
                }
                else -> mSongTag.visibility = View.GONE
            }
        }
    }
}